### PR TITLE
SparseLinear SparseJoinTable DenseToSparse

### PR DESCRIPTION
**Pull request by qiuxin2012**
## What changes were proposed in this pull request?
SparseLinear SparseJoinTable DenseToSparse

## How was this patch tested?
unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/DenseToSparse.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Convert DenseTensor to SparseTensor.
+ * @param ev$1
+ * @param ev
+ * @tparam T The numeric type in the criterion, usually which are [[Float]] or [[Double]]
+ */
+class DenseToSparse[T: ClassTag](implicit ev: TensorNumeric[T]) extends TensorModule[T] {
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    output = Tensor.sparse(input)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/DenseToSparse.scala]: input sparseTensor?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      initWeight: Tensor[T] = null,
+      initBias: Tensor[T] = null,
+      initGradWeight: Tensor[T] = null,
+      initGradBias: Tensor[T] = null)(implicit ev: TensorNumeric[T]) extends Linear[T](
+  inputSize, outputSize, withBias, wRegularizer, bRegularizer,
+  initWeight, initBias, initGradWeight, initGradBias) {
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 2,
+      "SparseLinear: " + ErrorInfo.constrainInputAsVectorOrBatch)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala]: check input type

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    if (withBias) output.addr(ev.one, addBuffer, bias)
+    output
+  }
+
+  // just backward a part of the gradOutput.
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 2,
+      "SparseLinear: " + ErrorInfo.constrainInputAsVectorOrBatch)
+    if (backwardStart >= 0 && backwardLength > 0) {
+      // TODO: _input to dense
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala]: unuse

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+      addBuffer.resize(Array(nFrame)).fill(ev.one)
+    }
+
+    SparseTensorBLAS.coomm(ev.one, input, weight.t, ev.zero, output)
+    if (withBias) output.addr(ev.one, addBuffer, bias)
+    output
+  }
+
+  // just backward a part of the gradOutput.
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala]: add comment

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      val _inputSize = Array(input.size(1), backwardLength)
+      val _weight = weight.narrow(2, backwardStart, backwardLength)
+
+      val nElement = gradInput.nElement()
+      gradInput.resize(_inputSize)
+      if (nElement != gradInput.nElement()) {
+        gradInput.zero()
+      }
+
+      gradInput.addmm(ev.fromType[Int](0), ev.fromType[Int](1), gradOutput, _weight)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala]: ev.one, ev.zero

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    output.resize(t)
+    if (output.nElement() != nElement) {
+      output.zero()
+    }
+
+    if (addBuffer.nElement() != nFrame) {
+      addBuffer.resize(Array(nFrame)).fill(ev.one)
+    }
+
+    SparseTensorBLAS.coomm(ev.one, input, weight.t, ev.zero, output)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SparseLinear.scala]: change to addmm

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 7]
```diff
@@ -634,6 +653,12 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
       nInputDims)
   }
 
+  def createSparseJoinTable(dimension: Int,
+                      nInputDims: Int)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: delete nInputDims

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/SparseTensor.scala [hunk 8]
```diff
@@ -991,6 +984,186 @@ private[tensor] class SparseTensor[@specialized(Float, Double) T: ClassTag](
...
+  }
+
+  /**
+   * Concatenate a sequence of SparseTensor of 1-dim to 2-dim SparseTensor.
+   *
+   * @param tensors a sequence of tensors
+   * @param res the resulted 2-dim SparseTensor
+   * @return res
+   */
+  private def concat[T: ClassTag](
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/SparseTensor.scala]: what?

**Comment by qiuxin2012** [comment]
jenkins passed

**Pull request merged**